src/engine.rs

```diff
@@ -80,7 +81,8 @@ impl Engine for PbftEngine {
             service,
             &mut pbft_state.write(),
         );
-
+        // block buffer for storing out of sequence blocks sent by validator
+        let mut block_buffer: HashMap<u64, Block> = HashMap::new();
```

Comment:
This should be added to `PbftState`, not passed down as an additional argument.

Additionally, it doesn't make sense to index the buffer by the block's number, since multiple blocks can exist at the same block number. I would probably replace this with a `HashSet` or a `Vec`.

Reply:
@ltseeley Since the main purpose of this is to facilitate the catchup process, the hypothesis was that there will only be one block at a block number that is committed. And the committed block will have the seal (thats why the seal check later on). We have seen some interesting issues, such as when one node was catching up and for some reason another node was stalled, the other node might keep proposing a block at an incorrect seq. This typically does not impact the regular cycle (since the PBFT seq_num would have advanced). But we have seen issues during catchup, where it stalls the node catching up if it receives such a block. 

Again, it will probably not hurt to have a hashSet and allow multiple blocks at the same blocknumber. But it may not make sense to keep blocks that are not yet committed from an efficiency standpoint. The Seal check further on is helping determine this. But is it possible that a  block that is not yet committed comes in with a seal? (In which case we may need to take the approach you recommend...)

Reply:
@itseeley Thanks for your reply . We have started working on your suggestion of including  the block_buffer in the PbftState .I wanted to know , What is your take on the @rajaramk suggestion ?
Do you think that the using the HashSet or a Vec wont have a significant impact on efficiency ?

Reply:
@Itseeley @vaporos  . I tried adding the block_buffer to the PbftState . But Faced the following error
erde::Serialize` is not satisfied\n\nthe trait `state::_IMPL_DESERIALIZE_FOR_PbftPhase::_serde::Serialize` is not implemented for `sawtooth_sdk::consensus::engine::Block

I guess the the trait is not implemented in the crate . I think we need to add it to crate and push it if we want to the add the block buffer to the PBFTState .  Please let me know if you think it is the correct way?